### PR TITLE
[TradingMode] Add DCA

### PR DESCRIPTION
**Pull request by Herklos**
Screenshot from 2021-04-24 13-07-28

With a 10s interval and 50$ config

### CODE REVIEW

File: Trading/Mode/dca_trading_mode/config/DCATradingMode_schema.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+  "type": "object",
+  "title": "DCATradingMode",
+  "properties": {
+    "amount_to_buy_in_reference_market": {
+        "title": "The amount of dollars (or unit of reference market) to buy on each transaction.",
+        "type": "number",
+        "multipleOf": 1.0,
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/config/DCATradingMode_schema.json:8]: think we should remove this multipleof condition to allow buy very low amounts (in btc for example)

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import octobot_trading.enums as trading_enums
+import octobot_trading.constants as trading_constants
+import octobot_trading.util as trading_util
+import octobot_trading.errors as trading_errors
+import octobot_trading.personal_data as trading_personal_data
+
+
+class DCATradingModeConsumer(trading_modes.AbstractTradingModeConsumer):
+    AMOUNT_TO_BUY_IN_REF_MARKET = "amount_to_buy_in_reference_market"
+    ORDER_PRICE_DISTANCE = 0.005
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/dca_trading.py:30]: 0.5% ? I think something like 0.1% would be better to unsure it's taken as maker but is not missed

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            self.logger.error(f"Failed to create order : {e}. Order: "
+                              f"{current_order.get_string_info() if current_order else None}")
+            self.logger.exception(e, False)
+            return []
+
+    async def can_create_order(self, symbol, state):
+        can_create_order_result = await super().can_create_order(symbol, state)
+        if not can_create_order_result:
+            currency, market = symbol_util.split_symbol(symbol)
+            self.logger.error(f"Can't create order : not enough balance. Please deposit more {market}.")
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/dca_trading.py:79]:
```suggestion
            self.logger.error(f"Can't create order : not enough balance. Please get more {market}.")
```

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        return can_create_order_result
+
+
+class DCATradingModeProducer(trading_modes.AbstractTradingModeProducer):
+    MINUTES_BEFORE_NEXT_BUY = "minutes_before_next_buy"
+    
+    def __init__(self, channel, config, trading_mode, exchange_manager):
+        super().__init__(channel, config, trading_mode, exchange_manager)
+        self.task = None
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/dca_trading.py:88]: we should cancel this task on stop()

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        # call orders creation from consumers
+        await self.submit_trading_evaluation(cryptocurrency=cryptocurrency,
+                                             symbol=symbol,
+                                             time_frame=None,
+                                             final_note=None,
+                                             state=self.state)
+
+        # send_notification
+        if not self.exchange_manager.is_backtesting:
+            await self._send_alert_notification(None)
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/dca_trading.py:107]:
```suggestion
            await self._send_alert_notification(symbol)
```

File: Trading/Mode/dca_trading_mode/dca_trading.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                self.logger.error(f"An error happened during DCA task : {e}")
+
+    async def start(self) -> None:
+        self.task = await asyncio.create_task(self.dca_task())
+
+    async def _send_alert_notification(self, symbol):
+        try:
+            import octobot_services.api as services_api
+            import octobot_services.enums as services_enum
+            title = f"OCTOBOT DCA ALERT : #{symbol}"
```
Comment (GuillaumeDSM) [inline Trading/Mode/dca_trading_mode/dca_trading.py:129]: I think we should use something else than "ALERT" here: "alert" can be scarry to users while this is an informative message, maybe "something like":
title = f"DCA trigger for #{symbol}"
(I also don't think we need "OctoBot", notifications are always in an OctoBot context already)

**Review by GuillaumeDSM — COMMENTED** [review]
great work, clear and useful !

File: Trading/Mode/dca_trading_mode/config/DCATradingMode_schema.json (hunk 1 above)
Comment (Herklos) [reply Trading/Mode/dca_trading_mode/config/DCATradingMode_schema.json:8]: Right :+1:

File: Trading/Mode/dca_trading_mode/dca_trading.py (hunk 2 above)
Comment (Herklos) [reply Trading/Mode/dca_trading_mode/dca_trading.py:30]: That's also right :+1:

**Review by GuillaumeDSM — APPROVED** [review]

**Pull request merged**